primitives/core/src/bandersnatch.rs

```diff
@@ -0,0 +1,1005 @@
+				self.secret.sign_thin_vrf(data.transcript.clone(), ios.as_slice());
+
+			let mut sign_bytes = [0; SIGNATURE_SERIALIZED_LEN];
+			signature
+				.signature
+				.serialize_compressed(sign_bytes.as_mut_slice())
+				.expect("serialization can't fail");
+
+			let outputs: Vec<_> = signature.preoutputs.into_iter().map(VrfOutput).collect();
+			let outputs = VrfIosVec::truncate_from(outputs);
+			VrfSignature { signature: Signature(sign_bytes), vrf_outputs: outputs }
+		}
+
+		/// Generate an arbitrary number of bytes from the given `context` and VRF `input`.
+		pub fn make_bytes<const N: usize>(
+			&self,
+			context: &'static [u8],
+			input: &VrfInput,
+		) -> [u8; N] {
+			let transcript = Transcript::new_labeled(context);
+			let inout = self.secret.clone().0.vrf_inout(input.0.clone());
+			inout.vrf_output_bytes(transcript)
+		}
+	}
+
+	impl Public {
+		fn vrf_verify_gen<const N: usize>(
+			&self,
+			data: &VrfSignData,
+			signature: &VrfSignature,
+		) -> bool {
+			let Ok(public) = PublicKey::deserialize_compressed(self.as_slice()) else {
+				return false
+			};
+
+			let Ok(preouts) = signature
+				.vrf_outputs
+				.iter()
+				.map(|o| o.0.clone())
+				.collect::<arrayvec::ArrayVec<bandersnatch_vrfs::VrfPreOut, N>>()
+				.into_inner()
+			else {
+				return false
+			};
+
+			// Deserialize only the proof, the rest has already been deserialized
+			// This is another hack used because backend signature type is generic over
+			// the number of ios.
+			let Ok(signature) =
+				ThinVrfSignature::<0>::deserialize_compressed(signature.signature.as_ref())
+					.map(|s| s.signature)
+			else {
+				return false
+			};
+			let signature = ThinVrfSignature { signature, preoutputs: preouts };
+
+			let inputs = data.vrf_inputs.iter().map(|i| i.0.clone());
+
+			signature.verify_thin_vrf(data.transcript.clone(), inputs, &public).is_ok()
+		}
+	}
+
+	impl VrfOutput {
+		/// Generate an arbitrary number of bytes from the given `context` and VRF `input`.
+		pub fn make_bytes<const N: usize>(
+			&self,
+			context: &'static [u8],
+			input: &VrfInput,
+		) -> [u8; N] {
+			let transcript = Transcript::new_labeled(context);
+			let inout =
+				bandersnatch_vrfs::VrfInOut { input: input.0.clone(), preoutput: self.0.clone() };
+			inout.vrf_output_bytes(transcript)
+		}
+	}
+}
+
+/// Bandersnatch Ring-VRF types and operations.
+pub mod ring_vrf {
+	use super::{vrf::*, *};
+	pub use bandersnatch_vrfs::ring::{RingProof, RingProver, RingVerifier, KZG};
+	use bandersnatch_vrfs::{CanonicalDeserialize, PedersenVrfSignature, PublicKey};
+
+	/// Context used to produce ring signatures.
+	#[derive(Clone)]
+	pub struct RingVrfContext(KZG);
+
+	impl RingVrfContext {
+		/// Build an dummy instance used for testing purposes.
+		pub fn new_testing() -> Self {
+			Self(KZG::testing_kzg_setup([0; 32], RING_MAX_SIZE))
+		}
+
+		/// Get the keyset size.
+		/// TODO @swasilyev: shouldn't this be equal to the domain_size used at init time?
```

Comment:
Should be domain_size - (253 + 3 + 1), where 253 is the Bandersnatch scalar bit size, 3 cells are used to make the construction ZK, and 1 cell is used to start the summation from.

See https://github.com/w3f/ring-proof/blob/0e948f3c28cbacecdd3020403c4841c0eb339213/ring/src/piop/params.rs#L32
and
https://github.com/w3f/ring-proof/blob/0e948f3c28cbacecdd3020403c4841c0eb339213/common/src/domain.rs#L62

Reply:
Got it. Thanks